flare/mgp/mapxb.py

```diff
@@ -0,0 +1,655 @@
+            self.build_map(GP)
+
+    def build_bond_struc(self, species_list):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def get_arrays(self, atom_env):
+        raise NotImplementedError("need to be implemented in child class")
+
+    def build_map_container(self, bounds):
+        '''
+        construct an empty spline container without coefficients.
+        '''
+
+        self.maps = []
+        for spc in self.spc:
+            m = self.singlexbody((self.grid_num, bounds, spc,
+                                  self.map_force, self.svd_rank, self.mean_only,
+                                  self.load_grid, self.lower_bound_relax,
+                                  self.n_cpus, self.n_sample))
+            self.maps.append(m)
+
+
+    def build_map(self, GP):
+        '''
+        generate/load grids and get spline coefficients
+        '''
+
+        self.kernel_info = get_kernel_term(GP, self.kernel_name)
+
+        for m in self.maps:
+            m.build_map(GP)
+
+
+    def predict(self, atom_env, mean_only):
+
+        min_dist = atom_env.bond_array_2[0][0]
+        lower_bound = np.max(self.maps[0].bounds[0][0])
+        assert min_dist >= lower_bound,\
+                f'The minimal distance {min_dist:.3f} is below the mgp lower bound {lower_bound:.3f}'
+
+        if self.mean_only:  # if not build mapping for var
+            mean_only = True
+
+        force_kernel, en_kernel, _, cutoffs, hyps, hyps_mask = self.kernel_info
+
+        args = from_mask_to_args(hyps, cutoffs, hyps_mask)
+
+        kern = 0
+        if not mean_only:
+            if self.map_force:
+                kern = np.zeros(3)
+                for d in range(3):
+                    kern[d] = force_kernel(atom_env, atom_env, d+1, d+1, *args)
+            else:
+                kern = en_kernel(atom_env, atom_env, *args)
+
+        spcs, comp_r, comp_xyz = self.get_arrays(atom_env)
+
+        # predict for each species
+        f_spcs = np.zeros(3)
+        vir_spcs = np.zeros(6)
+        v_spcs = np.zeros(3) if self.map_force else 0
+        e_spcs = 0
+        for i, spc in enumerate(spcs):
+            lengths = np.array(comp_r[i])
+            xyzs = np.array(comp_xyz[i])
+            map_ind = self.find_map_index(spc)
+
+            f, vir, v, e = self.maps[map_ind].predict(lengths, xyzs,
+                self.map_force, mean_only)
+            f_spcs += f
+            vir_spcs += vir
+            v_spcs += v
+            e_spcs += e
+
+        return f_spcs, vir_spcs, kern, v_spcs, e_spcs
+
+    def as_dict(self) -> dict:
+        """
+        Dictionary representation of the MGP model.
+        """
+
+        out_dict = deepcopy(dict(vars(self)))
+
+        # Uncertainty mappings currently not serializable;
+        if not self.mean_only:
+            out_dict['mean_only'] = True
+
+        # save the kernel name instead of callables
+        kernel, ek, efk, cutoffs, hyps, hyps_mask = self.kernel_info
```

Comment:
this block of code is the same as line 108. It should be isolated out as a function. And add a "kernel_info" argument or something equivalent to the init function